pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,334 @@
+
+	klog.InfoS("Starting external resizer for modify volume", "controller", ctrl.name)
+	defer klog.InfoS("Shutting down external resizer", "controller", ctrl.name)
+
+	stopCh := ctx.Done()
+	informersSyncd := []cache.InformerSynced{ctrl.pvInformer.HasSynced, ctrl.pvcInformer.HasSynced}
+	informersSyncd = append(informersSyncd, ctrl.vacInformer.HasSynced)
+
+	if !cache.WaitForCacheSync(stopCh, informersSyncd...) {
+		klog.ErrorS(nil, "Cannot sync pod, pv, pvc or vac caches")
+		return
+	}
+
+	// Cache all the InProgress/Infeasible PVCs as Uncertain for ModifyVolume
+	err := ctrl.initUncertainPVCs()
+	if err != nil {
+		klog.ErrorS(err, "Failed to initialize uncertain pvcs")
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(ctrl.sync, 0, stopCh)
+	}
+
+	<-stopCh
+}
+
+// sync is the main worker to sync PVCs.
+func (ctrl *modifyController) sync() {
+	key, quit := ctrl.claimQueue.Get()
+	if quit {
+		return
+	}
+	defer ctrl.claimQueue.Done(key)
+
+	if err := ctrl.syncPVC(key.(string)); err != nil {
+		// Put PVC back to the queue so that we can retry later.
+		klog.ErrorS(err, "Error syncing PVC")
+		ctrl.claimQueue.AddRateLimited(key)
+	} else {
+		ctrl.claimQueue.Forget(key)
+	}
+}
+
+// syncPVC checks if a pvc requests resizing, and execute the resize operation if requested.
+func (ctrl *modifyController) syncPVC(key string) error {
+	klog.V(4).InfoS("Started PVC processing for modify controller", "key", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("getting namespace and name from key %s failed: %v", key, err)
+	}
+
+	pvcObject, exists, err := ctrl.pvcInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("getting PVC %s/%s failed: %v", namespace, name, err)
+	}
+
+	if !exists {
+		klog.V(3).InfoS("PVC is deleted or does not exist", "PVC", klog.KRef(namespace, name))
+		return nil
+	}
+
+	pvc, ok := pvcObject.(*v1.PersistentVolumeClaim)
+	if !ok {
+		return fmt.Errorf("expected PVC got: %v", pvcObject)
+	}
+
+	if pvc.Spec.VolumeName == "" {
+		klog.V(4).InfoS("PV bound to PVC is not created yet", "PVC", klog.KObj(pvc))
+		return nil
+	}
+
+	volumeObj, exists, err := ctrl.pvInformer.GetStore().GetByKey(pvc.Spec.VolumeName)
+	if err != nil {
+		return fmt.Errorf("Get PV %q of pvc %q failed: %v", pvc.Spec.VolumeName, klog.KObj(pvc), err)
+	}
+	if !exists {
+		klog.InfoS("PV bound to PVC not found", "PV", pvc.Spec.VolumeName, "PVC", klog.KObj(pvc))
+		return nil
+	}
+
+	pv, ok := volumeObj.(*v1.PersistentVolume)
+	if !ok {
+		return fmt.Errorf("expected volume but got %+v", volumeObj)
+	}
+
+	// Only trigger modify volume if the following conditions are met
+	// 1. Feature gate is enabled
+	// 2. Non empty vac name
+	// 3. PVC is in Bound state
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeAttributesClass) {
+		vacName := pvc.Spec.VolumeAttributesClassName
+		if vacName != nil && *vacName != "" && pvc.Status.Phase == v1.ClaimBound {
```

Comment:
Why does a PV has to have `vacName`? Does this mean that this feature will not work for existing PV/PVCs which were created before this feature existed?

Reply:
This is when someone changed pvc.Spec.VolumeAttributesClassName and we trigger a modify, and yes it has to have a vacName, this is not about existing pv?

Reply:
So, what happens when `vac` is set for the very first time. Will the PV have `.Spec.VolumeAttributeClassName` populated?

Reply:
oops. this is pvc. I was thinking pv.